src/crystal/event_loop/io_uring.cr

```diff
@@ -0,0 +1,483 @@
+  end
+
+  # socket interface, see Crystal::EventLoop::Socket
+
+  def read(socket : ::Socket, slice : Bytes) : Int32
+    async_rw(LibC::IORING_OP_READ, socket, slice, socket.@read_timeout) do |errno|
+      case errno
+      when Errno::ECANCELED
+        raise IO::TimeoutError.new("Read timed out")
+      else
+        raise IO::Error.from_os_error("read", errno, target: socket)
+      end
+    end
+  end
+
+  def wait_readable(socket : ::Socket) : Nil
+    async_poll(socket, LibC::POLLIN | LibC::POLLRDHUP, socket.@read_timeout) { "Read timed out" }
+  end
+
+  def write(socket : ::Socket, slice : Bytes) : Int32
+    async_rw(LibC::IORING_OP_WRITE, socket, slice, socket.@write_timeout) do |errno|
+      case errno
+      when Errno::ECANCELED
+        raise IO::TimeoutError.new("Write timed out")
+      else
+        raise IO::Error.from_os_error("write", errno, target: socket)
+      end
+    end
+  end
+
+  def wait_writable(socket : ::Socket) : Nil
+    async_poll(socket, LibC::POLLOUT, socket.@write_timeout) { "Write timed out" }
+  end
+
+  def accept(socket : ::Socket) : ::Socket::Handle?
+    ret = async(LibC::IORING_OP_ACCEPT, socket.@read_timeout) do |sqe|
+      sqe.value.fd = socket.fd
+      sqe.value.sflags.accept_flags = LibC::SOCK_CLOEXEC
+    end
+    return ret unless ret < 0
+
+    if ret == -LibC::ECANCELED
+      raise IO::TimeoutError.new("Accept timed out")
+    elsif !socket.closed?
+      raise ::Socket::Error.from_os_error("accept", Errno.new(-ret))
+    end
+  end
+
+  def connect(socket : ::Socket, address : ::Socket::Addrinfo | ::Socket::Address, timeout : Time::Span?) : IO::Error?
+    sockaddr = address.to_unsafe # OPTIMIZE: #to_unsafe allocates (not needed)
+    addrlen = address.size
+
+    ret = async(LibC::IORING_OP_CONNECT, timeout) do |sqe|
+      sqe.value.fd = socket.fd
+      sqe.value.addr = sockaddr.address.to_u64!
+      sqe.value.u1.off = addrlen.to_u64!
+    end
+    return if ret == 0
+
+    if ret == -LibC::ECANCELED
+      IO::TimeoutError.new("Connect timed out")
+    elsif ret != -LibC::EISCONN
+      ::Socket::ConnectError.from_os_error("connect", Errno.new(-ret))
+    end
+  end
+
+  def send_to(socket : ::Socket, slice : Bytes, address : ::Socket::Address) : Int32
+    sockaddr = address.to_unsafe # OPTIMIZE: #to_unsafe allocates (not needed)
+    addrlen = address.size
+
+    if @@supports_sendto
+      ret = async(LibC::IORING_OP_SEND) do |sqe|
+        sqe.value.fd = socket.fd
+        sqe.value.addr = slice.to_unsafe.address.to_u64!
+        sqe.value.len = slice.size.to_u64!
+        sqe.value.u1.addr2 = sockaddr.address.to_u64!
+        sqe.value.addr_len[0] = addrlen.to_u16!
+      end
+      return ret unless ret < 0
+      raise ::Socket::Error.from_os_error("Error sending datagram to #{address}", Errno.new(-ret)) unless ret == -LibC::EINVAL
+      @@supports_sendto = false
+    end
+
+    ret = LibC.sendto(socket.fd, slice.to_unsafe.as(Void*), slice.size, 0, sockaddr, addrlen)
```

Comment:
It is possible to fall back to sendmsg op, but perhaps not worth the effort. My code for that is 
```
 vec = LibC::IOVec.new(base: slice.to_unsafe, len: slice.size)
      hdr = LibC::MsgHeader.new(
        name: addr.to_unsafe.as(LibC::SockaddrStorage*),
        namelen: LibC::SocklenT.new(sizeof(LibC::SockaddrStorage)),
        iov: pointerof(vec),
        iovlen: 1
      )

      get_sqe.sendmsg(socket, pointerof(hdr), user_data: userdata(scheduler))
 [.. submit and wait ..  ]
```
(the code is older than the sendto support in send..

Reply:
Oh, that's interesting.